### PR TITLE
Create Audio_visualizer_DSP.py

### PR DESCRIPTION
**Pull request by MortazaHassani**
Added audio visualizer for DSP

### Describe your change:

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: audio_filters/Audio_visualizer_DSP.py [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+import pyaudio
+from scipy.fftpack import fft
+from pynput.keyboard import Listener, Key
+
+
+class AudioStream:
+    """stream audio from input source (mic) and continuously
+    plot (bar) based on audio spectrum from waveform data"""
+
+    def __init__(self, num, symbol):
```
[the remark this answers is not in the document]
Comment (MortazaHassani) [reply audio_filters/Audio_visualizer_DSP.py:18]: def __init__(self, num, symbol) -> None:

File: audio_filters/Audio_visualizer_DSP.py [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+            "colorIndex_x": 100,
+            "colorIndex_y": 100,
+            "colorIndex_z": 255,
+        }
+
+        # graph init
+        self.graph = None
+
+    @staticmethod
+    def set_gradient_brush():
```
[the remark this answers is not in the document]
Comment (MortazaHassani) [reply audio_filters/Audio_visualizer_DSP.py:64]: QtGui.QBrush(grad)

File: audio_filters/Audio_visualizer_DSP.py [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+        grad.setColorAt(0.38, pg.mkColor("#FFFF00"))
+        grad.setColorAt(0.52, pg.mkColor("#00FF00"))
+        grad.setColorAt(0.66, pg.mkColor("#0000FF"))
+        grad.setColorAt(0.80, pg.mkColor("#4B0082"))
+        grad.setColorAt(0.94, pg.mkColor("#9400D3"))
+        grad.setCoordinateMode(QtGui.QGradient.ObjectMode)
+        return QtGui.QBrush(grad)
+
+    # Bar Graph
+    def set_plotdata_1(self, name, data_x, data_y):
```
[the remark this answers is not in the document]
Comment (MortazaHassani) [reply audio_filters/Audio_visualizer_DSP.py:78]: def set_plotdata_1(self, name, data_x, data_y) -> None:

File: audio_filters/Audio_visualizer_DSP.py [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+            self.traces.add(name)
+            # initial setup of bar plot
+            brush = self.set_gradient_brush()
+            self.graph = pg.BarGraphItem(
+                x=data_x, height=data_y, width=50, brush=brush, pen=(0, 0, 0)
+            )
+        self.audio_plot.addItem(self.graph)
+
+    # Scatter Plot Graph
+    def set_plotdata_2(self, name, data_x, data_y):
```
[the remark this answers is not in the document]
Comment (MortazaHassani) [reply audio_filters/Audio_visualizer_DSP.py:94]: def set_plotdata_2(self, name, data_x, data_y) -> None:

name -> string
data_x -> array
data_y -> array

File: audio_filters/Audio_visualizer_DSP.py [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+            )
+            self.graph = pg.PlotCurveItem(x=data_x, y=data_y, pen=pen1,)
+        else:
+            self.traces.add(name)
+            # initial setup of curve plot
+            self.graph = pg.PlotCurveItem(x=data_x, y=data_y, pen="r",)
+        self.audio_plot.addItem(self.graph)
+
+    # Line Graph
+    def set_plotdata_4(self, name, data_x, data_y):
```
[the remark this answers is not in the document]
Comment (MortazaHassani) [reply audio_filters/Audio_visualizer_DSP.py:144]: def set_plotdata_4(self, name, data_x, data_y) -> None:
name -> string
data_x -> array
data_y -> array

File: audio_filters/Audio_visualizer_DSP.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (MortazaHassani) [reply audio_filters/Audio_visualizer_DSP.py:78]: def set_plotdata_1(self, name, data_x, data_y) -> None:

name -> string
data_x -> array
data_y -> array

File: audio_filters/Audio_visualizer_DSP.py [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+                y=data_y,
+                pen=None,
+                symbol=self.symbol,
+                size=30,
+                brush=(100, 100, 255, 100),
+            )
+        self.audio_plot.addItem(self.graph)
+
+    # Curve Graph
+    def set_plotdata_3(self, name, data_x, data_y):
```
[the remark this answers is not in the document]
Comment (MortazaHassani) [reply audio_filters/Audio_visualizer_DSP.py:117]: def set_plotdata_3(self, name, data_x, data_y) -> None:

name -> string
data_x -> array
data_y -> array

File: audio_filters/Audio_visualizer_DSP.py [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+            )
+            self.graph.setData(data_x, data_y, pen=pen1, shadowPen="#19070B")
+        else:
+            pen1 = pg.mkPen(color=(250, 0, 0), width=15, style=QtCore.Qt.DashLine)
+            self.traces.add(name)
+            # initial setup of curve plot
+            self.graph = pg.PlotCurveItem(x=data_x, y=data_y, pen=pen1, shadowPen="r",)
+        self.audio_plot.addItem(self.graph)
+
+    def update(self):
```
[the remark this answers is not in the document]
Comment (MortazaHassani) [reply audio_filters/Audio_visualizer_DSP.py:170]: def update(self) -> None:

File: audio_filters/Audio_visualizer_DSP.py [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+            2: self.set_plotdata_2,
+            3: self.set_plotdata_3,
+            4: self.set_plotdata_4,
+        }
+
+        plot_data.get(self.number)(
+            name="spectrum", data_x=self.f, data_y=self.calculate_data()
+        )
+
+    def calculate_data(self):
```
[the remark this answers is not in the document]
Comment (MortazaHassani) [reply audio_filters/Audio_visualizer_DSP.py:183]: def calculate_data(self) -> None:

File: audio_filters/Audio_visualizer_DSP.py [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+            else:
+                (
+                    self.a["colorIndex_x"],
+                    self.a["colorIndex_y"],
+                    self.a["colorIndex_z"],
+                ) = color_index.get(key)
+        except Exception as error:
+            print(error)
+
+    def color_index_control(self):
```
[the remark this answers is not in the document]
Comment (MortazaHassani) [reply audio_filters/Audio_visualizer_DSP.py:248]: def color_index_control(self) -> None:

File: audio_filters/Audio_visualizer_DSP.py [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+        for color_rgb in ("colorIndex_x", "colorIndex_y", "colorIndex_z"):
+            if self.a[color_rgb] >= 255:
+                self.a[color_rgb] = 0
+
+    def animation(self):
```
[the remark this answers is not in the document]
Comment (MortazaHassani) [reply audio_filters/Audio_visualizer_DSP.py:253]: def animation(self) -> None:

File: audio_filters/Audio_visualizer_DSP.py [hunk 11]
```diff
@@ -0,0 +1,298 @@
...
+        """call self.start and self.update for continuous
+        output application"""
+        timer = QtCore.QTimer()
+        timer.timeout.connect(self.update)
+        timer.start(20)
+        self.start()
+
+
+def int_to_symbol(symbol):
```
[the remark this answers is not in the document]
Comment (MortazaHassani) [reply audio_filters/Audio_visualizer_DSP.py:262]: def int_to_symbol(symbol) -> None:

File: audio_filters/Audio_visualizer_DSP.py [hunk 12]
```diff
@@ -0,0 +1,312 @@
...
+from pynput.keyboard import Key, Listener
+from pyqtgraph.Qt import QtCore, QtGui
+from scipy.fftpack import fft
+
+
+class AudioStream:
+    """stream audio from input source (mic) and continuously
+    plot (bar) based on audio spectrum from waveform data"""
+
+    def __init__(self, num, symbol):
```
[the remark this answers is not in the document]
Comment (MortazaHassani) [reply audio_filters/Audio_visualizer_DSP.py:18]: def __init__(self, num, symbol) -> None:

File: audio_filters/Audio_visualizer_DSP.py [hunk 13]
```diff
@@ -0,0 +1,312 @@
...
+        grad.setColorAt(0.38, pg.mkColor("#FFFF00"))
+        grad.setColorAt(0.52, pg.mkColor("#00FF00"))
+        grad.setColorAt(0.66, pg.mkColor("#0000FF"))
+        grad.setColorAt(0.80, pg.mkColor("#4B0082"))
+        grad.setColorAt(0.94, pg.mkColor("#9400D3"))
+        grad.setCoordinateMode(QtGui.QGradient.ObjectMode)
+        return QtGui.QBrush(grad)
+
+    # Bar Graph
+    def set_plotdata_1(self, name, data_x, data_y):
```
[the remark this answers is not in the document]
Comment (MortazaHassani) [reply audio_filters/Audio_visualizer_DSP.py:78]: def set_plotdata_1(self, name, data_x, data_y) -> None:
name  -> None
data_x -> None
data_y -> None

File: audio_filters/Audio_visualizer_DSP.py [hunk 14]
```diff
@@ -0,0 +1,312 @@
...
+            "colorIndex_x": 100,
+            "colorIndex_y": 100,
+            "colorIndex_z": 255,
+        }
+
+        # graph init
+        self.graph = None
+
+    @staticmethod
+    def set_gradient_brush():
```
[the remark this answers is not in the document]
Comment (MortazaHassani) [reply audio_filters/Audio_visualizer_DSP.py:64]: def set_gradient_brush() -> None:

File: audio_filters/Audio_visualizer_DSP.py [hunk 15]
```diff
@@ -0,0 +1,312 @@
...
+            self.traces.add(name)
+            # initial setup of bar plot
+            brush = self.set_gradient_brush()
+            self.graph = pg.BarGraphItem(
+                x=data_x, height=data_y, width=50, brush=brush, pen=(0, 0, 0)
+            )
+        self.audio_plot.addItem(self.graph)
+
+    # Scatter Plot Graph
+    def set_plotdata_2(self, name, data_x, data_y):
```
[the remark this answers is not in the document]
Comment (MortazaHassani) [reply audio_filters/Audio_visualizer_DSP.py:94]: def set_plotdata_2(self, name, data_x, data_y) -> None:
name -> None
data_x -> None
data_y -> None

File: audio_filters/Audio_visualizer_DSP.py [hunk 16]
```diff
@@ -0,0 +1,312 @@
...
+        sp_data = (
+            np.abs(sp_data[0 : int(self.CHUNK)])  # slice: slice first half of our fft
+            * 2
+            / (256 * self.CHUNK)
+        )  # rescale: mult 2, div amp waveform and no. freq in your spectrum
+        sp_data[sp_data <= 0.001] = 0
+        return sp_data
+
+    @staticmethod
+    def start():
```
[the remark this answers is not in the document]
Comment (MortazaHassani) [reply audio_filters/Audio_visualizer_DSP.py:204]: def start() -> None:

File: audio_filters/Audio_visualizer_DSP.py [hunk 17]
```diff
@@ -0,0 +1,312 @@
...
+        """start application"""
+        if (sys.flags.interactive != 1) or not hasattr(QtCore, "PYQT_VERSION"):
+            QtGui.QApplication.instance().exec_()
+
+    def change_color(self, key):
```
[the remark this answers is not in the document]
Comment (MortazaHassani) [reply audio_filters/Audio_visualizer_DSP.py:209]: def change_color(self, key) -> None:

File: audio_filters/Audio_visualizer_DSP.py [hunk 18]
```diff
@@ -0,0 +1,312 @@
...
+            else:
+                (
+                    self.a["colorIndex_x"],
+                    self.a["colorIndex_y"],
+                    self.a["colorIndex_z"],
+                ) = color_index.get(key)
+        except Exception as error:
+            print(error)
+
+    def color_index_control(self):
```
[the remark this answers is not in the document]
Comment (MortazaHassani) [reply audio_filters/Audio_visualizer_DSP.py:248]: def color_index_control(self) -> None:

File: audio_filters/Audio_visualizer_DSP.py [hunk 19]
```diff
@@ -0,0 +1,312 @@
...
+        for color_rgb in ("colorIndex_x", "colorIndex_y", "colorIndex_z"):
+            if self.a[color_rgb] >= 255:
+                self.a[color_rgb] = 0
+
+    def animation(self):
```
[the remark this answers is not in the document]
Comment (MortazaHassani) [reply audio_filters/Audio_visualizer_DSP.py:253]: def animation(self) -> None:

File: audio_filters/Audio_visualizer_DSP.py [hunk 20]
```diff
@@ -0,0 +1,312 @@
...
+        """call self.start and self.update for continuous
+        output application"""
+        timer = QtCore.QTimer()
+        timer.timeout.connect(self.update)
+        timer.start(20)
+        self.start()
+
+
+def int_to_symbol(symbol):
```
[the remark this answers is not in the document]
Comment (MortazaHassani) [reply audio_filters/Audio_visualizer_DSP.py:262]: def int_to_symbol(symbol) -> None:

File: audio_filters/Audio_visualizer_DSP.py (hunk 12 above)
Comment (cclauss) [inline audio_filters/Audio_visualizer_DSP.py:18]: It is a bot therefore it is persistent.  If you say that this is resolved but it is not then the bot will keep saying it is not.
```suggestion
    def __init__(self, num: int, symbol: int) -> None:
```

File: audio_filters/Audio_visualizer_DSP.py [hunk 21]
```diff
@@ -0,0 +1,312 @@
...
+    print("Choose and type number.")
+    print(line_break)
+    print("1: Bar Graph")
+    print("2: Scatter Graph")
+    print("3: Curve Graph")
+    print("4: Line Graph")
+    print("5: quit")
+    print(line_break)
+    symbol = "o"  # default symbol to instantiate AudioStream class
+    number_input = input("Graph Type: ")
```
Comment (cclauss) [inline audio_filters/Audio_visualizer_DSP.py:283]: See CONTRIBUTING.md
```suggestion
    number_input = input("Graph Type: ").strip()
```

**Comment by cclauss** [comment]
[CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md) was not read.  Please do not open new pull requests until November 1st. 2022.

**Pull request closed without merging**